samples/bluetooth/direct_test_mode/src/transport/hci_uart.c

```diff
@@ -0,0 +1,302 @@
+static const struct device *hci_uart_dev = DEVICE_DT_GET(DTM_UART);
+
+NET_BUF_POOL_DEFINE(hci_tx_buf, QUEUE_COUNT, QUEUE_SIZE, 0, NULL);
+static K_FIFO_DEFINE(hci_tx_queue);
+
+NET_BUF_POOL_DEFINE(hci_rx_buf, QUEUE_COUNT, QUEUE_SIZE, sizeof(uint8_t), NULL);
+
+enum h4_state {
+	S_TYPE,
+	S_HEADER,
+	S_PAYLOAD
+};
+
+static hci_uart_read_cb dtm_hci_put;
+
+static int hci_hdr_len(uint8_t type)
+{
+	switch (type) {
+	case H4_TYPE_CMD:
+		return sizeof(struct bt_hci_cmd_hdr);
+
+	case H4_TYPE_ACL:
+		return sizeof(struct bt_hci_acl_hdr);
+
+	case H4_TYPE_EVT:
+		return sizeof(struct bt_hci_evt_hdr);
+
+	case H4_TYPE_ISO:
+		return sizeof(struct bt_hci_iso_hdr);
+
+	default:
+		return 0;
+	}
+}
+
+static int hci_pld_len(uint8_t type, uint8_t *hdr)
+{
+	switch (type) {
+	case H4_TYPE_CMD:
+		return ((struct bt_hci_cmd_hdr *)hdr)->param_len;
+
+	case H4_TYPE_ACL:
+		return sys_le16_to_cpu(((struct bt_hci_acl_hdr *)hdr)->len);
+
+	case H4_TYPE_ISO:
+		return sys_le16_to_cpu(((struct bt_hci_iso_hdr *)hdr)->len);
+
+	default:
+		return 0;
+	}
+}
+
+static bool h4_rx_type(uint8_t type)
+{
+	return ((type == H4_TYPE_CMD) | (type == H4_TYPE_ACL) | (type == H4_TYPE_ISO));
+}
+
+static size_t buf_read(struct net_buf *buf, const uint8_t *src, size_t src_len, size_t req_len)
+{
+	size_t len;
+
+	if (req_len > src_len) {
+		len = src_len;
+	} else {
+		len = req_len;
+	}
+
+	if (net_buf_tailroom(buf) < len) {
+		__ASSERT_NO_MSG(false);
+	}
+	net_buf_add_mem(buf, src, len);
+
+	return len;
+}
+
+static void h4_read(const uint8_t *data, size_t offset, size_t len)
+{
+	static enum h4_state state = S_TYPE;
+	static struct net_buf *buf;
+	static uint8_t type;
+	static int rem;
+	size_t read;
+
+	while (len > 0) {
+		switch (state) {
+		case S_TYPE:
+			type = data[offset];
+			offset += sizeof(type);
+			len -= sizeof(type);
+
+			if (h4_rx_type(type)) {
+				rem = hci_hdr_len(type);
+				buf = net_buf_alloc(&hci_rx_buf, K_NO_WAIT);
+				if (!buf) {
+					__ASSERT_NO_MSG(false);
+					/* Out of buffers condition */
+				}
+
+				*(uint8_t *)net_buf_user_data(buf) = type;
+				state = S_HEADER;
+			} else {
+				/* TODO: Sync failure */
+			}
+			break;
+
+		case S_HEADER:
+			read = buf_read(buf, &data[offset], len, rem);
+			offset += read;
+			len -= read;
+			rem -= read;
+			if (rem == 0) {
+				rem = hci_pld_len(type, buf->data);
+				state = S_PAYLOAD;
+				if (rem == 0) {
+					if (dtm_hci_put) {
+						dtm_hci_put(buf);
+					} else {
+						LOG_ERR("Callback dtm_hci_put is not assigned.");
+					}
+					state = S_TYPE;
+				}
+			}
+			break;
+
+		case S_PAYLOAD:
+			read = buf_read(buf, &data[offset], len, rem);
+			offset += read;
+			len -= read;
+			rem -= read;
+			if (rem == 0) {
+				if (dtm_hci_put) {
+					dtm_hci_put(buf);
+				} else {
+					LOG_ERR("Callback dtm_hci_put is not assigned.");
+				}
+				state = S_TYPE;
+			}
+			break;
+
+		default:
+			state = S_TYPE;
+			break;
+		}
+	}
+}
+
+static uint8_t *uart_buf(void)
+{
+	static int cur;
```

Comment:
Signed integer overflow is undefined behavior in C
```suggestion
	static uint32_t cur;
```